implement aligned realloc, use memalign, allow use on embedded

corecode opened the pull request:
Quick 2 commits.  Suggest fast forward instead of merge.

corecode committed:
make realloc honor alignment

This uses the same approach as in std: alloc, copy, and dealloc

daniel5151 commented:
Sure, these look good to me.

Any more coming, or do you want me to publish these as a 1.0.5?

daniel5151 commented on Cargo.toml:
```diff
@@ -12,4 +12,7 @@ repository  = "https://github.com/daniel5151/libc_alloc"
 categories = ["development-tools::ffi", "embedded", "no-std"]
 
+[features]
+libc = []
```
mind adding a comment explaining what this feature does?

daniel5151 reviewed (changes requested):
Ah, though, some documentation would be good...

corecode commented:
Yes, I'll add some docs.

Maybe we should default to "libc" if windows isn't set?  After all, the crate name is "libc_alloc".

daniel5151 commented on src/lib.rs:
```diff
@@ -16,10 +16,11 @@
 use core::ptr;
 
-#[cfg(target_family = "unix")]
+#[cfg(any(target_family = "unix", feature = "libc"))]
```
Now that I think about it... maybe this should just be

```rust
#[cfg(not(target_family = "windows"))]
```

and avoid introducing the new feature entirely?

corecode replied:
I prefer your suggestion.

daniel5151 reviewed (changes requested):
Left a suggestion that might solve the issue you identified without requiring a new feature + additional docs.

Let me know what you think

corecode committed:
use memalign instead of posix_memalign

A commonly used libc for embedded systems, newlib, does not provide
posix_memalign, so use plain memalign instead.

corecode committed:
default to using libc functions unless we are on windows

This allows us to use libc for embedded systems where rustc does not
set target_family.

corecode commented:
What do you think about this?

daniel5151 commented:
Is there a reason you're switching from `posix_memalign` to just `memalign`?

If anything, we should be duplicating this logic from the stdlib?

https://doc.rust-lang.org/src/std/sys/unix/alloc.rs.html#55-102

corecode commented:
Like I wrote in the commit, newlib does not provide posix_memalign.

While glibc's manpage states:
 > POSIX requires that memory obtained from posix_memalign() can be freed using free(3).  Some systems provide no way to reclaim memory allocated  with  memalign() or valloc() (because one can pass to free(3) only a pointer obtained from malloc(3), while, for example, memalign() would call malloc(3) and then align the obtained value).  The glibc implementation allows memory obtained from any of these functions to be reclaimed with free(3).

I think that those systems are very rare, and they won't have posix_memalign either.

daniel5151 commented:
Ahh, didn't catch that extra context in the commit messages, my bad 😅

Sure, in that case, this all looks good to me.
I'll go ahead and land these commits.

Thanks for the PR!

* * *

Like I asked above: Any more changes coming, or would you like me to to publish a new 1.0.5 release?

corecode commented:
No changes from me at the moment, good to release.

Thanks!

daniel5151 commented:
Cool, I've published 1.0.5 to crates.io.

Cheers!
